File: mongo/src/main/java/org/enodeframework/mongo/MongoPublishedVersionStore.java
package org.enodeframework.mongo;

import com.mongodb.MongoWriteException;
import com.mongodb.client.model.Filters;
import com.mongodb.client.model.Updates;
import com.mongodb.client.result.InsertOneResult;
import com.mongodb.client.result.UpdateResult;
import com.mongodb.reactivestreams.client.MongoClient;
import org.bson.Document;
import org.bson.conversions.Bson;
import org.enodeframework.common.exception.EnodeRuntimeException;
import org.enodeframework.common.exception.IORuntimeException;
import org.enodeframework.eventing.IPublishedVersionStore;
import org.reactivestreams.Subscriber;
import org.reactivestreams.Subscription;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.SQLException;
import java.util.Date;
import java.util.concurrent.CompletableFuture;

/**
 * @author anruence@gmail.com
 */
public class MongoPublishedVersionStore implements IPublishedVersionStore {

    private static final Logger logger = LoggerFactory.getLogger(MongoPublishedVersionStore.class);

    private final MongoClient mongoClient;

    private final int duplicateCode;

    private final String uniqueIndexName;

    private final MongoConfiguration configuration;

    public MongoPublishedVersionStore(MongoClient mongoClient) {
        this(mongoClient, new MongoConfiguration());
    }

    public MongoPublishedVersionStore(MongoClient mongoClient, MongoConfiguration configuration) {
        this.configuration = configuration;
        this.mongoClient = mongoClient;
        this.uniqueIndexName = configuration.getPublishedVersionUniqueIndexName();
        this.duplicateCode = configuration.getDuplicateCode();
    }

    @Override
    public CompletableFuture<Integer> updatePublishedVersionAsync(String processorName, String aggregateRootTypeName, String aggregateRootId, int publishedVersion) {
        CompletableFuture<Integer> future = new CompletableFuture<>();
        boolean isInsert = publishedVersion == 1;
        if (isInsert) {
            Document document = new Document();
            document.put("processorName", processorName);
            document.put("aggregateRootTypeName", aggregateRootTypeName);
            document.put("aggregateRootId", aggregateRootId);
            document.put("version", 1);
            document.put("gmtCreate", new Date());
            mongoClient.getDatabase(configuration.getDatabaseName()).getCollection(configuration.getPublishedVersionCollectionName())
                    .insertOne(document).subscribe(new Subscriber<InsertOneResult>() {
                @Override
                public void onSubscribe(Subscription s) {
                    s.request(1);
                }

                @Override
                public void onNext(InsertOneResult x) {
                    future.complete(x.wasAcknowledged() ? 1 : 0);
                }

                @Override
                public void onError(Throwable t) {
                    future.completeExceptionally(t);
                }

                @Override
                public void onComplete() {
                    future.complete(1);
                }
            });
        } else {
            Bson filter = Filters.and(
                    Filters.eq("version", publishedVersion - 1),
                    Filters.eq("processorName", processorName),
                    Filters.eq("aggregateRootId", aggregateRootId)
            );
            Bson update = Updates.combine(
                    Updates.set("version", publishedVersion),
                    Updates.set("gmtCreate", new Date())
            );
            mongoClient.getDatabase(configuration.getDatabaseName()).getCollection(configuration.getPublishedVersionCollectionName())
                    .updateOne(filter, update).subscribe(new Subscriber<UpdateResult>() {
                private int updated;

                @Override
                public void onSubscribe(Subscription s) {
                    s.request(1);
                }

                @Override
                public void onNext(UpdateResult x) {
                    updated = ((int) x.getModifiedCount());
                }

                @Override
                public void onError(Throwable t) {
                    future.completeExceptionally(t);
                }

                @Override
                public void onComplete() {
                    future.complete(updated);
                }
            });
        }

        return future.exceptionally(throwable -> {
            if (throwable instanceof MongoWriteException) {
                MongoWriteException ex = (MongoWriteException) throwable;
                if (isInsert && ex.getCode() == duplicateCode && ex.getMessage().contains(uniqueIndexName)) {
                    return 0;
                }
                logger.error("Insert or update aggregate published version has sql exception.", ex);
                throw new IORuntimeException(throwable);
            }
            logger.error("Insert or update aggregate published version has unknown exception.", throwable);
            throw new EnodeRuntimeException(throwable);

        });
    }

    @Override
    public CompletableFuture<Integer> getPublishedVersionAsync(String processorName, String aggregateRootTypeName, String aggregateRootId) {
        CompletableFuture<Integer> future = new CompletableFuture<>();
        Bson updateFilter = Filters.and(
                Filters.eq("processorName", processorName),
                Filters.eq("aggregateRootId", aggregateRootId)
        );
        mongoClient.getDatabase(configuration.getDatabaseName()).getCollection(configuration.getPublishedVersionCollectionName()).find(updateFilter).subscribe(new Subscriber<Document>() {
            private Integer version = 0;

            @Override
            public void onSubscribe(Subscription s) {
                s.request(1);
            }

            @Override
            public void onNext(Document document) {
                version = document.getInteger("version");
                future.complete(version);
            }

            @Override
            public void onError(Throwable t) {
                future.completeExceptionally(t);
            }

            @Override
            public void onComplete() {
                future.complete(version);
            }
        });
        return future.exceptionally(throwable -> {
            if (throwable instanceof SQLException) {
                SQLException ex = (SQLException) throwable;
                logger.error("Get aggregate published version has sql exception.", ex);
                throw new IORuntimeException(throwable);
            }
            logger.error("Get aggregate published version has unknown exception.", throwable);
            throw new EnodeRuntimeException(throwable);
        });
    }
}

File: mongo/src/main/java/org/enodeframework/mongo/MongoPublishedVersionStore.kt
package org.enodeframework.mongo

import com.google.common.collect.Lists
import com.mongodb.client.model.Updates
import io.vertx.core.json.JsonObject
import io.vertx.ext.mongo.MongoClient
import org.enodeframework.eventing.EventStoreConfiguration
import org.enodeframework.eventing.PublishedVersionStore
import org.enodeframework.mongo.handler.MongoAddPublishedVersionHandler
import org.enodeframework.mongo.handler.MongoFindPublishedVersionHandler
import org.enodeframework.mongo.handler.MongoUpdatePublishedVersionHandler
import java.util.*
import java.util.concurrent.CompletableFuture

/**
 * @author anruence@gmail.com
 */
open class MongoPublishedVersionStore(
    private val mongoClient: MongoClient, private val options: EventStoreConfiguration
) : PublishedVersionStore {
    override fun updatePublishedVersionAsync(
        processorName: String, aggregateRootTypeName: String, aggregateRootId: String, publishedVersion: Int
    ): CompletableFuture<Int> {
        val insert = publishedVersion == 1
        if (insert) {
            return insertAsync(processorName, aggregateRootTypeName, aggregateRootId, publishedVersion)
        }
        return updateAsync(processorName, aggregateRootTypeName, aggregateRootId, publishedVersion)
    }

    private fun insertAsync(
        processorName: String, aggregateRootTypeName: String, aggregateRootId: String, publishedVersion: Int
    ): CompletableFuture<Int> {
        val document = JsonObject()
        document.put("processorName", processorName)
        document.put("aggregateRootTypeName", aggregateRootTypeName)
        document.put("aggregateRootId", aggregateRootId)
        document.put("version", 1)
        document.put("gmtCreate", Date().toInstant())
        val handler = MongoAddPublishedVersionHandler(
            "$processorName#$aggregateRootTypeName#$aggregateRootId#$publishedVersion", options.publishedUkName
        )
        mongoClient.insert(options.publishedTableName, document, handler)
        return handler.future
    }

    private fun updateAsync(
        processorName: String, aggregateRootTypeName: String, aggregateRootId: String, publishedVersion: Int
    ): CompletableFuture<Int> {
        val map1 = HashMap<String, Int>()
        val map2 = HashMap<String, String>()
        val map3 = HashMap<String, String>()
        map1["version"] = publishedVersion - 1
        map2["aggregateRootId"] = aggregateRootId
        map3["processorName"] = processorName
        val queryJson = JsonObject()
        val queryFilter = Lists.newArrayList(map1, map2, map3)
        queryJson.put("\$and", queryFilter)
        val update = Updates.combine(
            Updates.set("version", publishedVersion), Updates.set("gmtCreate", Date().toInstant())
        )
        val updateJson = JsonObject(update.toBsonDocument().toJson())
        val publishHandle = MongoUpdatePublishedVersionHandler(
            "$processorName#$aggregateRootTypeName#$aggregateRootId#$publishedVersion",
        )
        mongoClient.updateCollection(options.publishedTableName, queryJson, updateJson, publishHandle)
        return publishHandle.future
    }

    override fun getPublishedVersionAsync(
        processorName: String, aggregateRootTypeName: String, aggregateRootId: String
    ): CompletableFuture<Int> {
        val map1 = HashMap<String, String>()
        val map2 = HashMap<String, String>()
        map1["processorName"] = processorName
        map2["aggregateRootId"] = aggregateRootId
        val queryJson = JsonObject()
        val filter = Lists.newArrayList(map1, map2)
        queryJson.put("\$and", filter)
        val handler = MongoFindPublishedVersionHandler("$aggregateRootId#$processorName#$aggregateRootTypeName")
        mongoClient.findOne(options.publishedTableName, queryJson, null, handler)
        return handler.future
    }
}